libs/editor/egui_editor/src/ast.rs

```diff
@@ -0,0 +1,135 @@
+use crate::cursor_types::DocByteOffset;
+use crate::element::Element;
+use pulldown_cmark::{Event, OffsetIter, Options, Parser};
+use std::ops::Range;
+
+#[derive(Default, Debug)]
+pub struct Ast {
+    pub nodes: Vec<AstNode>,
+    pub root: usize,
+}
+
+#[derive(Default, Debug)]
+pub struct AstNode {
+    pub element: Element,
+    pub range: Range<DocByteOffset>,
+    pub children: Vec<usize>,
+}
+
+impl Ast {
+    pub fn parse(raw: &str) -> Self {
+        let mut options = Options::empty();
+        options.insert(Options::ENABLE_STRIKETHROUGH);
+        let parser = Parser::new_ext(raw, options);
+        let mut result = Self {
+            nodes: vec![AstNode::new(
+                Element::Document,
+                Range { start: DocByteOffset(0), end: DocByteOffset(raw.len()) },
+            )],
+            root: 0,
+        };
+        result.push_children(result.root, &mut parser.into_offset_iter(), raw);
+        result
+    }
+
+    fn push_children(&mut self, current_idx: usize, iter: &mut OffsetIter, raw: &str) {
+        while let Some((event, range)) = iter.next() {
+            let mut range =
+                Range { start: DocByteOffset(range.start), end: DocByteOffset(range.end) };
+            match event {
+                Event::Start(new_child) => {
+                    if let Some(new_child) = Element::from_tag(new_child) {
+                        if new_child == Element::Item {
+                            range.start -= Self::look_back_whitespace(raw, range.start);
+                        }
+                        if new_child == Element::CodeBlock {
+                            range.end +=
+                                Self::capture_codeblock_newline(raw, range.start, range.end);
+                        }
+
+                        let new_child_idx =
+                            self.push_child(current_idx, AstNode::new(new_child, range));
+                        self.push_children(new_child_idx, iter, raw);
+                    }
+                }
+                Event::Code(_) => {
+                    self.push_child(current_idx, AstNode::new(Element::InlineCode, range));
+                }
+                Event::End(done) => {
+                    if let Some(done) = Element::from_tag(done) {
+                        if done == self.nodes[current_idx].element {
+                            break;
+                        }
+                    }
+                }
+                _ => {} // todo: there are some interesting events ignored (rules, tables, etc)
```

Comment:
I would just make this a normal comment above this line and drop the 'todo' part, but just resolve if you want to leave as is.

Reply:
resolving